lib/AccessStaticPropertyWithinModelContextRule.php

```diff
@@ -0,0 +1,80 @@
+<?php
+
+declare(strict_types=1);
+
+namespace SlamPhpStan;
+
+use PhpParser\Node;
+use PhpParser\Node\Expr\StaticPropertyFetch;
+use PHPStan\Analyser\Scope;
+use PHPStan\Broker\Broker;
+use PHPStan\Rules\Rule;
+
+final class AccessStaticPropertyWithinModelContextRule implements Rule
+{
+    /**
+     * @var Broker
+     */
+    private $broker;
+
+    /**
+     * @var string
+     */
+    private $modelBaseClassOrInterface;
+
+    /**
+     * @var string
+     */
+    private $singletonAccessor;
+
+    public function __construct(Broker $broker, string $modelBaseClassOrInterface, string $singletonAccessor)
+    {
+        $this->broker                    = $broker;
+        $this->modelBaseClassOrInterface = $modelBaseClassOrInterface;
+        $this->singletonAccessor         = $singletonAccessor;
+    }
+
+    public function getNodeType(): string
+    {
+        return StaticPropertyFetch::class;
+    }
+
+    /**
+     * @param \PhpParser\Node\Expr\StaticPropertyFetch $node
+     * @param \PHPStan\Analyser\Scope                  $scope
+     *
+     * @return string[] errors
+     */
+    public function processNode(Node $node, Scope $scope): array
+    {
+        if (! $node->class instanceof Node\Name || ! $node->name instanceof Node\VarLikeIdentifier) {
+            return [];
+        }
+
+        if (! $scope->isInClass()) {
+            return [];
+        }
+
+        $classReflection = $scope->getClassReflection();
+        if (null === $classReflection || ! $classReflection->isSubclassOf($this->modelBaseClassOrInterface)) {
```

Comment:
`$classReflection` can no longer be null here - I recommend using an extension from PHPStan, register it like this: https://github.com/phpstan/phpstan/blob/master/build/phpstan.neon#L53-L59

Reply:
I'm sorry may I ask you why that extension isn't globally active? I mean, in which cases it would be necessary to be turned off?

Reply:
This extension is relevant only for other PHPStan extension developers. It does not make sense to run it on a project that just uses PHPStan and does not developer extension by itself. That's why it's in `PHPStan\Internal`.